### PR TITLE
jg/faster data loader

### PR DESCRIPTION
**Pull request by james-golden-arcadia**
The data loader was updated so that the training is faster. 

## PR checklist
- [ x ] No previous issue
- [ x ] Updated the data loader to make it higher throughput (at least on gpu)
- [ x ] Added test/test_sampler.py; new data loader shows a meaningful speedup for gene_expression.ipynb (for 24 GB VRAM, 4:00 to 1:40 for training)
- [ x ] No new dependencies
- [ x ] No new functionality
- [ x ] gene_expression.ipynb is stale and will be updated in a future PR

Note that a few files were added for an estimator for optimal batch size, but they did not work and were removed.

### CODE REVIEW

File: src/glass_box_umap/parametric_umap/core.py [hunk 1]
```diff
@@ -23,6 +23,21 @@
...
+    "16-mixed",
+    "bf16-true",
+    "bf16-mixed",
+    "32-true",
+    "64-true",
+    "64",
+    "32",
+    "16",
+    "bf16",
+]
```
Comment (ekiefl) [inline src/glass_box_umap/parametric_umap/core.py:39]: These are unrelated to the data loader and they aren't safe yet. The
description says "No new functionality", but this adds a public dataclass field
and more importantly, the knob is broken/untested. The codebase currently
supports a fixed datatype and offering tunable datatypes to the users demands a
full review of how this impacts the API surface, for example serialization
round trips, behavior/guards against certain datatypes on certain devices, etc.
Much work is to be done if we are to support this, so please remove `precision`
and `TrainerPrecision` from this PR.

File: src/glass_box_umap/parametric_umap/core.py [hunk 2]
```diff
@@ -155,6 +171,20 @@ def _build_model(self, input_dims: tuple[int, ...]) -> UMAPLightningModule:
 
         return model
 
+    def _build_training_graph(self, X: NDArray[np.float32]):
```
Comment (ekiefl) [inline src/glass_box_umap/parametric_umap/core.py:174]: Please add a return annotation, which should be `csr_matrix`

File: src/glass_box_umap/parametric_umap/core.py [hunk 3]
```diff
@@ -224,13 +255,7 @@ def fit(self, X: NDArray[np.floating] | Tensor) -> Self:
...
             # higher-dim input (e.g. images for ConvEncoder) for graph
             # construction only; UMAPDataset still receives the original X.
-            graph = get_umap_graph(
-                X.reshape(X.shape[0], -1) if X.ndim > 2 else X,
-                n_neighbors=self.n_neighbors,
-                metric=self.metric,
-                random_state=self.random_state,
-                quiet=self.quiet,
-            )
+            graph = self._build_training_graph(X)
```
Comment (ekiefl) [inline src/glass_box_umap/parametric_umap/core.py:258]: The three-line comment above explaining why NNDescent needs 2D input stayed in `fit()` while the reshape it describes moved into the method. Maybe move the 3-line comment into the docstring of the new `_build_training_graph`?

File: src/glass_box_umap/__init__.py [hunk 4]
```diff
@@ -14,4 +14,5 @@ def __getattr__(name: str) -> str:
 __all__ = [
     "GlassBoxUMAP",
     "ParametricUMAP",
+    "recommend_batch_size",
```
Comment (ekiefl) [inline src/glass_box_umap/__init__.py:17]: This object doesn't exist, so this should be deleted. Currently `from glass_box_umap import *` now raises an `AttributeError`.

File: src/glass_box_umap/parametric_umap/lightning/data.py [hunk 5]
```diff
@@ -1,88 +1,118 @@
...
+        batch_sampler = SplitAndMergeWeightedBatchSampler(
+            weights=torch.as_tensor(self.dataset.edge_weights, dtype=torch.double),
+            num_samples=len(self.dataset),
+            batch_size=self.batch_size,
+        )
+        return DataLoader(
+            dataset=self.dataset,
+            # DataLoader accepts tensor batches at runtime and forwards them to
+            # Dataset.__getitems__, but PyTorch's annotation permits lists only.
+            batch_sampler=cast(Any, batch_sampler),
```
Comment (ekiefl) [inline src/glass_box_umap/parametric_umap/lightning/data.py:108]: This cast is covering something real: `batch_sampler` is declared `Sampler[list] | Iterable[list] | None`, but `_MapDatasetFetcher.fetch` is itself unannotated and hands the batch straight to `__getitems__`, so torch's runtime contract is broader than its annotation. The comment above the cast describes this accurately. My suggestion is only about how to implement the type check skip:

```python
batch_sampler=batch_sampler,  # pyright: ignore[reportArgumentType]
```

A stale `cast(Any, ...)` is undetectable, since casting to `Any` is never redundant. Applying this and the point below also removes `from typing import Any, cast` from the module entirely.

File: src/glass_box_umap/parametric_umap/lightning/data.py [hunk 6]
```diff
@@ -1,88 +1,118 @@
...
+            num_workers=self.num_workers,
+            persistent_workers=self.num_workers > 0,
+            collate_fn=cast(Any, _identity_collate),
```
Comment (ekiefl) [inline src/glass_box_umap/parametric_umap/lightning/data.py:111]: `cast(Any, _identity_collate)` is only needed because the function is annotated
`tuple[Tensor, Tensor] -> tuple[Tensor, Tensor]` while torch declares
`_collate_fn_t = Callable[[list[_T]], Any]`. Annotating it as a real identity
function typechecks clean with no cast:

```python
_T = TypeVar("_T")

def _identity_collate(batch: _T) -> _T:
    return batch
```

It stays a module-level function, so it still pickles for `num_workers > 0`.

**Review by ekiefl — CHANGES_REQUESTED** [review]
The core change is good, speedup is crazy fast! I reproduced it independently: iterating one
epoch of the loader alone, on 20,000 samples with 256 features and roughly 300,000 edges at batch
size 10,000, went from 0.834 s on `main` to 0.037 s on this branch, about a 22-fold reduction on
the data-loading path. That's consistent with the 4:00 to 1:40 end-to-end figure in the
description. The sampling is also bit-identical for a given seed rather than merely equivalent in
distribution, which `test_tensor_batch_sampler_matches_scalar_draws` asserts directly, so this is
purely a throughput change and not a change in training geometry.
